### DISCUSSION TITLE
Cannot get page to work off of center of image instead of top left of image OR why a circle of 3 images is not offset adequately

### DISCUSSION THREAD

**Discussion by Exaskryz** [ImageMagick/ImageMagick#6636] [Help] [answered]
### ImageMagick version

6.9.11-60

### Operating system, version and so on

Ubuntu 22.04 LTS

### Description

I'm looking here: https://www.imagemagick.org/Usage/layers/#layer_thumbnails at the second example that shows thumbnails arranged in a circle.

The documentation says 

>The position calculated is of the top-left corner of the image (not its center, though that is a simple adjustment),

But it fails to ever tell us how to make that adjustment. I have tried my guess at using +w/2 and +h/2 in the given expression but that is not making any changes to my final image.

Here are some example results of what I'm getting. Using 4 or 10 images, looks great. The odd numbers of images are not perfect, but acceptable. When it's 2 images, that's okay too, as another image is going to obstruct the center and so the off-by-a-pixel alignment is not a concern. (Could be managed by making sure the rotations are done on an image with odd height probably if anyone really cared.) But the problem is 3 images. 

https://imgur.com/a/dBIfAFV

What I am thinking is I need the displacement of images to be calculated from the middle of each image. But I could absolutely be wrong on that and it's not changing the results. In which case, is there an order of operations I should change and possibly avoiding repage to keep working with the same home coordinates?

Ultimately I could just make a special case where when count = 3 to shift the spokes by some measure, but I would really like to have more flexible code that won't need those exceptions.

_Additional thoughts. The testouput2.png image which is the spokes prior to shedding is meeting the angle of 120 degrees it looks like. But they aren't making it to the corners, as the roses are arranged, and if I understand geometry, for them to meet at the corners we would be producing a right triangle and need that to be 90 degrees. So something should change where either the roses are not offset by 90/135/135 or the spokes are, which could be part of my exception._

```#!bin/bash

count=$1
duplications=$((count-1))
ratio=240

convert rose: -duplicate "$duplications" -set page '+%[fx:('$ratio'*cos((t/n)*2*pi)+100)]+%[fx:('$ratio'*sin((t/n)*2*pi)+100)]' -background none -layers merge +repage testoutput_alpha.png

alpha_x=$(identify -ping -format '%w' testoutput_alpha.png)
alpha_y=$(identify -ping -format '%h' testoutput_alpha.png)

alpha_short_x=$((alpha_x/2))

if [ $alpha_y -lt $alpha_short_x ] 
then 
	convert testoutput_alpha.png -background none -gravity Center -extent "$alpha_x"x"$alpha_x" testoutput_alpha.png 
	echo did this, it was a big horizontal rectangle. This should only ever happen when only 2 images are submitted for the "circle"
fi

full_x=$(identify -ping -format '%w' testoutput_alpha.png) #yes this is redundant from the above because I cropped out a step in making my condensed version. Apologies. This normally refers to a different intermediary image.
full_y=$(identify -ping -format '%h' testoutput_alpha.png)
half_x=$(($full_x/2))
half_y=$(($full_y/2))
custom_x=$(($full_x*93/200))

#This is creating a large white circle to be a mask to create a final result of basically the spokes like a bicycle wheel.

convert -size "$full_x"x"$full_y" xc:transparent testpremask.png
convert testpremask.png -fill white -stroke white -draw "translate "$half_x","$half_y" circle 0,0 "$custom_x",0" testmask.png

#This creates a set of spokes from center of image out to the edges.

convert -size "$full_x"x"$full_y" xc:transparent testoutput2.png
i=0
while [ $i -lt $count ]
do
	convert testoutput2.png -stroke white -strokewidth 10 -draw "line "$half_x","$half_y" "$full_x","$half_y"" -distort SRT ""$half_x","$half_y" %[fx:360/"$count"]" testoutput2.png
	((i++))
done

#Next shrink the spokes by a little bit so they are bounded by an invisible circle, essentially.

composite -compose Dst_In testmask.png testoutput2.png testresult.png

#quick create a pretty background

convert -size "$full_x"x"$full_y" gradient:crimson-orchid3 testgradient.png

#mash it all together

convert testgradient.png testresult.png testoutput_alpha.png -flatten testfinal.png```

**Comment by snibgo** [comment]
Please re-post your script in a way that the forum software doesn't reformat it. I suggest putting a line of 3 backticks above and below the script.

I suggest upgrading to IM v7 if you can. V6 `convert` is now very old. V7 is much better.

**Reply by Exaskryz** [reply]
Is there a way to update Ubuntu's apt-get list then to do v7? All I did was apt-get install imagemagick or something out of the terminal, and that's what it gave me.

Might also ask if github can add a code button for multiline. I just tapped the code button in the post editor, the one that is an icon of < >. So had no idea about triple backtick. Will edit post for that.

**Comment by fmw42** [comment]
Image:

[images.zip](https://github.com/ImageMagick/ImageMagick/files/12578395/images.zip)


Here is the original code using radius to the top left.  

```
cd 
cd desktop/images
radius=80
im7 magick balloon.gif castle.gif ghost.gif medical.gif \
present.gif skull.gif snowman.gif terminal.gif tree.gif \
-set page "+%[fx:$radius*cos((t/n)*2*pi)]+%[fx:$radius*sin((t/n)*2*pi)]" \
-background none -layers merge +repage image_circle.png

```

![image_circle](https://github.com/ImageMagick/ImageMagick/assets/13458218/d86e5b84-240a-48aa-8fc8-dc6bb5164819)



Here is the new code using the radius to the center:

```
cd 
cd desktop/images
radius=80
im7 magick balloon.gif castle.gif ghost.gif medical.gif \
present.gif skull.gif snowman.gif terminal.gif tree.gif \
-set page "+%[fx:($radius+w/2)*cos((t/n)*2*pi)]+%[fx:($radius+h/2)*sin((t/n)*2*pi)]" \
-background none -layers merge +repage image_circle2.png

```

![image_circle2](https://github.com/ImageMagick/ImageMagick/assets/13458218/25ee5a2d-6064-421f-82d4-474162d8f5ca)


Does that do what you want?  If not, try

cd 
cd desktop/images
radius=80
im7 magick balloon.gif castle.gif ghost.gif medical.gif \
present.gif skull.gif snowman.gif terminal.gif tree.gif \
-set page "+%[fx:($radius+sqrt(w*w/4+h*h/4))*cos((t/n)*2*pi)]+%[fx:($radius+sqrt(w*w/4+h*h/4))*sin((t/n)*2*pi)]" \
-background none -layers merge +repage image_circle3.png

![image_circle3](https://github.com/ImageMagick/ImageMagick/assets/13458218/baed4b47-03ae-43c9-9c67-11344ca3f931)

**Reply by Exaskryz** [reply]
The first example and middle example seem to be what I was asking, but it's not quite getting me to the answer. They look the same. So I must be approaching the problem incorrectly - misdiagnosing it. Did you look at my imgur album to see what happens with spokes when it's 3 of them? The spokes generate a 120 degree angle between each other, but the images are not arranged at 120 degrees apart.

I went ahead and did the 3-item exception in my code and found out that the angles that seem to work best are 130/100/130 (Adds up to 360). Those make spokes that touch my images. (Well, maybe not the roses, but the ones I am using in production.)

So the question is then, why does the circle code not produce evenly spaced images? How could that be achieved? Will setting a larger canvas size to start fix that?

**Comment by GeeMack** [comment]
In a single ImageMagick command like this, you can create the spokes, arrange the overlays, and composite all of it onto a background. It's quite a different approach than your example, but might give you some alternative ideas. The command is in Windows syntax and uses IMv6 command syntax, but easily translates to *nix. 

Two places in the command you need to tell it how many spokes, and two places in the command you need to tell it how far from the outer edge you want to place the items. Obviously these can be sent into the command as variables. The spokes can be as many or few as you want, up until there are so many that they pile up on each other.

    magick convert -size 640x640 -background none ^
       xc:none ^
       -draw "stroke white stroke-width 10 line 320,320 320,64" ^
       -duplicate 6 ^
       -distort SRT "%[fx:t*360/n]" -flatten ^
       -write mpr:spokes +delete ^
       rose: ^
       -gravity center -extent 640x640 +gravity ^
       -duplicate 6 ^
       -distort SRT "%[fx:w/2],%[fx:h/2] 1 -%[fx:t*360/n] %[fx:w/2],64" ^
       -distort SRT "%[fx:t*360/n]" -flatten ^
       mpr:spokes gradient:lightgreen-lightblue -reverse -flatten result.png

That command run on Windows produces this result...

![result](https://github.com/ImageMagick/ImageMagick/assets/65475279/afa9dc4b-b130-4988-adae-e075e49b6eec)

The command starts by creating a single line on a transparent canvas, duplicates it to the desired number of spokes, rotates each as needed, and flattens them to a single image which is then saved in a memory register "mpr:spokes".

That spoke image is deleted from the active command, and the "rose:" image is read in. That image is placed in the center of a transparent canvas of the finished output size and duplicated to the required number.

Then a "-distort" operation moves each of those the proper distance from the center, and *also* rotates each -360/number_of_spokes degrees so they sort of unwind as they're put in their final locations in the next step...

The final "-distort" operation rotates each of those canvases to place the image at the proper location around the circumference. Then those images are flattened.

After that the "mpr:spokes" image is brought back into the command, we create a gradient background, and flatten all the parts to produce the output image.

**Reply by Exaskryz** [reply]
This is very promising and likely the code I had wanted to know how to write from the start. I had all my incremental steps really just to troubleshoot where my math had gone wrong and I might have eventually tried to consolidate it. (My real code ends up using 7 intermediary images. Very ugly, yes.) I am just off to bed now after work + football and will be happy to read over this. Thank you so much for explaining what looks to be every bit of the code and what it's doing!

**Comment by snibgo** [comment]
@Exaskryz: Perhaps you don't realise that the code you are using positions the top-left corner of each small image in the `-page` position. If using `-page`, you also need to take the dimensions of the small images into account.

See GeeMacks's solution.

**Reply by fmw42** [reply]
@GeeMack,  your code should use magick, not magick convert in IM 7.

**Reply by Exaskryz** [reply]
In the OP I mentioned trying to add w/2 and h/2. This was not in the published code, because what I was doing with w/2 and h/2 did not seem to make any changes, at least not what I was gunning for.

"I have tried my guess at using +w/2 and +h/2 in the given expression but that is not making any changes to my final image."

**Comment by fmw42** [comment, marked as answer]
**Here is a correction to my code above and a detailed explanation**

The old example from Anthony draws the image and places them at the top left corner according to the specified radius as follows:

```
cd 
cd desktop/images
radius=80
magick balloon.gif castle.gif ghost.gif medical.gif \
present.gif skull.gif snowman.gif terminal.gif tree.gif \
-set page "+%[fx:$radius*cos((t/n)*2*pi)]+%[fx:$radius*sin((t/n)*2*pi)]" \
-background none -layers merge +repage image_circle.png

```

![image_circle](https://github.com/ImageMagick/ImageMagick/assets/13458218/d9a9427a-3a0c-441d-9128-0b6fdc9d39b4)


If we want to specify the radius to the center, we still need to compute the placement according to the top left corner as required by -set page.  So we have to compute the center and then offset the center by half the dimensions of the images to compute the +X+Y offsets needed by -set page to the top left corner. (The top left corner X = center x - w/2 and top left corner Y = center y - h/2, where w, h are the dimensions of each image)

For 9 images, the code would be the following. Note that I have added a blue dot at the center of each image, placed those images and then post-process to add lines from the center of the output at 9 equal angles of the specified radius.  We see that the lines come close to the centers of the images. They are off slightly due to the non-equal dimensions of the output and placements of the centers.

```
cd 
cd desktop/images
radius=80
magick balloon.gif castle.gif ghost.gif medical.gif \
present.gif skull.gif snowman.gif terminal.gif tree.gif \
-fill blue -draw "rectangle %[fx:w/2-2],%[fx:h/2-2] %[fx:w/2+2],%[fx:h/2+2]" \
-set page "+%[fx:($radius)*cos((t/n)*2*pi)-w/2]+%[fx:($radius)*sin((t/n)*2*pi)-h/2]" \
-background none -layers merge +repage image_circle5_9.png

radius=80
for ((i=0; i<9; i++)); do
magick image_circle5_9.png \
-fill blue -draw "line %[fx:w/2],%[fx:h/2] %[fx:$radius*cos(($i/9)*2*pi)+w/2],%[fx:$radius*sin(($i/9)*2*pi)+h/2]" \
image_circle5_9.png 
done
magick image_circle5_9.png -compose copy -bordercolor red -border 1 image_circle5_9.png 

```

![image_circle5_9](https://github.com/ImageMagick/ImageMagick/assets/13458218/919f9301-13ac-4da1-8028-e1dce01d38b1)



Here is for 7 image:

```
cd 
cd desktop/images
radius=80
magick balloon.gif castle.gif ghost.gif medical.gif \
present.gif skull.gif snowman.gif \
-fill blue -draw "rectangle %[fx:w/2-2],%[fx:h/2-2] %[fx:w/2+2],%[fx:h/2+2]" \
-set page "+%[fx:($radius)*cos((t/n)*2*pi)-w/2]+%[fx:($radius)*sin((t/n)*2*pi)-h/2]" \
-background none -layers merge +repage image_circle5_7.png

radius=80
for ((i=0; i<7; i++)); do
magick image_circle5_7.png \
-fill blue -draw "line %[fx:w/2],%[fx:h/2] %[fx:$radius*cos(($i/7)*2*pi)+w/2],%[fx:$radius*sin(($i/7)*2*pi)+h/2]" \
image_circle5_7.png 
done
magick image_circle5_7.png -compose copy -bordercolor red -border 1 image_circle5_7.png 

```

![image_circle5_7](https://github.com/ImageMagick/ImageMagick/assets/13458218/623e22eb-6fda-4363-915d-0e8452a054fb)





Here is for 3 images:

```
cd 
cd desktop/images
radius=80
magick balloon.gif castle.gif ghost.gif \
-fill blue -draw "rectangle %[fx:w/2-2],%[fx:h/2-2] %[fx:w/2+2],%[fx:h/2+2]" \
-set page "+%[fx:($radius)*cos((t/n)*2*pi)-w/2]+%[fx:($radius)*sin((t/n)*2*pi)-h/2]" \
-background none -layers merge +repage image_circle5_3.png

radius=80
for ((i=0; i<3; i++)); do
magick image_circle5_3.png \
-fill blue -draw "line %[fx:w/2],%[fx:h/2] %[fx:$radius*cos(($i/3)*2*pi)+w/2],%[fx:$radius*sin(($i/3)*2*pi)+h/2]" \
image_circle5_3.png 
done
magick image_circle5_3.png -compose copy -bordercolor red -border 1 image_circle5_3.png 

```

![image_circle5_3](https://github.com/ImageMagick/ImageMagick/assets/13458218/597e0c15-acb6-4afe-9d6c-f98a454237d7)

Note that the equal angle lines do not end at the centers of the images.  This is due to the way the images are drawn at the top left corner in Anthony's code due to the use of -set page. This results in a non-square image.

One way to compensate for this is to add more invisible images. So adding 2 invisible images between each actual image,
results in:
 
```
cd 
cd desktop/images
radius=80
img_size="32x32"
magick -respect-parenthesis \
balloon.gif \( -size "$img_size" xc:none +duplicate \) castle.gif \( -size "$img_size" xc:none +duplicate \) ghost.gif \( -size "$img_size" xc:none +duplicate \)  \
-fill blue -draw "rectangle %[fx:w/2-2],%[fx:h/2-2] %[fx:w/2+2],%[fx:h/2+2]" \
-set page "+%[fx:($radius)*cos((t/n)*2*pi)-w/2]+%[fx:($radius)*sin((t/n)*2*pi)-h/2]" \
-background none -layers merge +repage image_circle5_3b.png

radius=80
for ((i=0; i<3; i++)); do
magick image_circle5_3b.png \
-fill blue -draw "line %[fx:w/2],%[fx:h/2] %[fx:$radius*cos(($i/3)*2*pi)+w/2],%[fx:$radius*sin(($i/3)*2*pi)+h/2]" \
image_circle5_3b.png 
done
magick image_circle5_3b.png -compose copy -bordercolor red -border 1 image_circle5_3b.png 

```

![image_circle5_3b](https://github.com/ImageMagick/ImageMagick/assets/13458218/2292bf76-e72f-4090-a27a-28891e15c73c)

If the center dots are removed, we get:

```
cd 
cd desktop/images
radius=80
img_size="32x32"
magick -respect-parenthesis \
balloon.gif \( -size "$img_size" xc:none +duplicate \) castle.gif \( -size "$img_size" xc:none +duplicate \) ghost.gif \( -size "$img_size" xc:none +duplicate \)  \
-set page "+%[fx:($radius)*cos((t/n)*2*pi)-w/2]+%[fx:($radius)*sin((t/n)*2*pi)-h/2]" \
-background none -layers merge +repage image_circle5_3c.png

radius=80
for ((i=0; i<3; i++)); do
magick image_circle5_3c.png \
-fill blue -draw "line %[fx:w/2],%[fx:h/2] %[fx:$radius*cos(($i/3)*2*pi)+w/2],%[fx:$radius*sin(($i/3)*2*pi)+h/2]" \
image_circle5_3c.png 
done
magick image_circle5_3c.png -compose copy -bordercolor red -border 1 image_circle5_3c.png 

```

![image_circle5_3c](https://github.com/ImageMagick/ImageMagick/assets/13458218/2ea18c93-9191-4c5b-8ef4-f4bcbf347da9)


Skipping the code to draw the lines, we get

```
cd 
cd desktop/images
radius=80
img_size="32x32"
magick -respect-parenthesis \
balloon.gif \( -size "$img_size" xc:none +duplicate \) castle.gif \( -size "$img_size" xc:none +duplicate \) ghost.gif \( -size "$img_size" xc:none +duplicate \)  \
-set page "+%[fx:($radius)*cos((t/n)*2*pi)-w/2]+%[fx:($radius)*sin((t/n)*2*pi)-h/2]" \
-background none -layers merge +repage image_circle5_3d.png

magick image_circle5_3d.png -compose copy -bordercolor red -border 1 image_circle5_3d.png 

```

![image_circle5_3d](https://github.com/ImageMagick/ImageMagick/assets/13458218/b477432d-ea11-405e-9c99-e483801547a4)


We can now get rid of the excess transparency on the left by adding -trim:

```
cd 
cd desktop/images
radius=80
img_size="32x32"
magick -respect-parenthesis \
balloon.gif \( -size "$img_size" xc:none +duplicate \) castle.gif \( -size "$img_size" xc:none +duplicate \) ghost.gif \( -size "$img_size" xc:none +duplicate \)  \
-set page "+%[fx:($radius)*cos((t/n)*2*pi)-w/2]+%[fx:($radius)*sin((t/n)*2*pi)-h/2]" \
-background none -layers merge -trim +repage image_circle5_3e.png

magick image_circle5_3e.png -compose copy -bordercolor red -border 1 image_circle5_3e.png 

```

![image_circle5_3e](https://github.com/ImageMagick/ImageMagick/assets/13458218/c2980ec6-c70f-49c4-a9dc-0b4bd53d1cf1)

However, the center of the result is not the center for the radial lines.

**Reply by Exaskryz** [reply]
This is very helpful and what I'll consider the answer. I'll do some more reading and cross reference with the docs. But it looks like because either I was using intermediary images, and then calculating my spokes based on the dimensions of the intermediary images, or possibly using "repage" will do trimming, that is how my centers become misaligned for the two purposes. 

GeeMack's answer may be the closest match to the code I desire to use. But this is the answer for me because it explains where my thinking and assumptions had gone wrong and helps me learn, rather than scrap what I have and forget about it. As I continue to improve and revise my code, I may find myself working toward GeeMack's example.

**Comment by fmw42** [comment]
Here is GeeMack's solution for 7 image in Unix syntax.  The only downside that I see is that one has to specify a size that is larger than needed and then -trim. It does not output a size that depends upon the radius and image sizes only automatically.

Here I specify a size of 250x250 (for -extent) for 7 images and radius 80.

```
cd 
cd desktop/images
radius=80
magick \
-background none \
balloon.gif castle.gif ghost.gif medical.gif \
present.gif skull.gif snowman.gif \
-gravity center -extent 250x250 +gravity \
-distort SRT "%[fx:w/2],%[fx:h/2] 1 -%[fx:t*360/n] %[fx:w/2],%[fx:h/2-$radius]" \
-distort SRT "%[fx:t*360/n]" -background white -flatten -trim +repage \
-bordercolor red -border 1 \
result.png
```

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/c080c138-e751-417a-a047-023addb813d5)
